src/sparseml/exporters/transforms/kv_cache/transforms_opt.py

```diff
@@ -12,122 +12,148 @@
-    |     |
-    |  Gather(model.decoder.embed_positions.weight)
-    ```
-
+    1. Adds `positions` as an input to the model
+    2. Adds `causal_mask` as an input to the model
+    2. Finds the node that initially creates the `embed_position.weight` tensor
+    3. Updates the node to use the positions input instead of
+       computing it from the Sub op
+    4. Finds the nodes that initially create the `causal_mask` tensors
+    5. Updates the nodes to use the causal_mask input instead of
+          computing it from the Expand op
+
+    :param model: model to update
+    :return: updated model
     """
 
+    POSITION_EMBEDDINGS_IDS_MATCHING_PATTERN = dict(
+        op_type="Sub",
+        children_ops=[["Add"]],
+    )
+    CAUSAL_MASK_MATCHING_PATTERN = dict(op_type="Expand", children_ops=[["Add"]])
+
     def transform(self, model: ModelProto) -> ModelProto:
         model = self.add_positions_input(model)
-        position_embeddings_node = self.find_embed_positions_gather_node(model)
-        model = self._update_position_embeddings_for_graph_input(
-            model, position_embeddings_node
+        model = self.add_causal_mask_input(model)
+
+        position_embeddings_nodes = self.find_nodes_by_pattern(
+            model, pattern=self.POSITION_EMBEDDINGS_IDS_MATCHING_PATTERN
         )
-        return model
+        if len(position_embeddings_nodes) != 1:
+            raise ValueError(
+                "Expected to find exactly one node matching "
+                f"the pattern {self.POSITION_EMBEDDINGS_IDS_MATCHING_PATTERN}, "
+                f"found {len(position_embeddings_nodes)}"
+            )
+        model = self.inject_positions(model, position_embeddings_nodes, "Add")
 
-    @classmethod
-    def find_embed_positions_gather_node(cls, model: ModelProto) -> NodeProto:
-        for node in model.graph.node:
-            if node.op_type != "Gather":
-                continue
-            if node.input[0] == _EMBED_POSITIONS_ID:
-                # found the embed_positions_gather_node
-                return node
-        raise RuntimeError(
-            f"Unable to find position embeddings gather node with id "
-            f"{_EMBED_POSITIONS_ID} in {cls.__name__}"
+        causal_mask_nodes = self.find_nodes_by_pattern(
+            model, pattern=self.CAUSAL_MASK_MATCHING_PATTERN
         )
+        model = self.inject_causal_mask(model, causal_mask_nodes, "Add")
+        model = self.adjust_causal_mask(model)
+        return model
+
+    def adjust_causal_mask(self, model: ModelProto) -> ModelProto:
+        """
+        Insert a `Where`` node after the causal mask input to change
+        the initial boolean mask, to a mask of floats expected by the model.
+
+
+        Transform:
+        ```
+        |       causal_mask
+        |            |
+        |   causal_mask_input_child
+        ```
+        to:
+        ```
+        |       causal_mask
+        |            |
+        |          Where
```

Comment:
why are we using a where instead of a cast to bool? can you check with runtime for what would work better if either are fine? additionally, seems like a cast would read better in onnx vs where which involves a condition...